### DISCUSSION TITLE
Seconds in exif gps values for ios mobile pics are always null

### DISCUSSION THREAD

**Discussion by Dyloneus** [dlemstra/Magick.NET#733] [Help] [answered]
I've run into a problem where exif gps info from ios devices. The [seconds] rational (from ExifTag.GPSLongitude array) is always 0.

It works ok with android and other images. Here's what i'm trying to do:

`internal static ExifGPSCoordinates GetGPSCoordinates(IExifProfile exifProfile)
        {
            ExifGPSCoordinates gpsCoordinates = null;
            var gpsLongitude = exifProfile.GetValue(ExifTag.GPSLongitude);
            var gpsLongitudeRef = exifProfile.GetValue(ExifTag.GPSLongitudeRef);
            var gpsLatitude = exifProfile.GetValue(ExifTag.GPSLatitude);
            var gpsLatitudeRef = exifProfile.GetValue(ExifTag.GPSLatitudeRef);

            if ((gpsLongitudeRef != null && gpsLongitude != null) &&
                (gpsLatitudeRef != null && gpsLatitude != null))
            {
                gpsCoordinates = new ExifGPSCoordinates
                {
                    Longitude = (decimal)GetCoordinates(gpsLongitudeRef.ToString(), gpsLongitude.Value),
                    Latitude = (decimal)GetCoordinates(gpsLatitudeRef.ToString(), gpsLatitude.Value)
                };
            }

            return gpsCoordinates;
        }

        private static double? GetCoordinates(string gpsRef, Rational[] rationals)
        {
            if (string.IsNullOrEmpty(gpsRef) && (rationals == null || rationals.Length == 0))
            {
                return null;
            }

            double degrees = rationals[0].Numerator / rationals[0].Denominator;
            double minutes = rationals[1].Numerator / rationals[1].Denominator;
            double seconds = rationals[2].Numerator / rationals[2].Denominator;

            double coordinate = degrees + (minutes / 60d) + (seconds / 3600d);
            if (gpsRef == "S" || gpsRef == "W")
                coordinate *= -1;
            return coordinate;
        }
`

Ive attached a sample ios image for testing. 

![APC_0331](https://user-images.githubusercontent.com/17324490/93327023-a5527b00-f811-11ea-80d7-dca5d6b2484f.jpg)

You can check the metadata on https://www.pic2map.com/ which shows the coordinates to be 45° 30' 42.30" N , 122° 39' 56.38" W.

However, in the code above, the value in the [seconds] rational is 0.

Is this a bug or should I be parsing GPSLongitude and GPSLatitude out differently?

Thanks in advance for your help

**Comment by dlemstra** [comment]
Just noticed that you are doing the conversion to `double` yourself. You probably want to do `rational.ToDouble()` instead. This is unrelated to your question.

**Comment by dlemstra** [comment, marked as answer]
I got the correct result with the following calucation:

```C#
var value = profile.GetValue(ExifTag.GPSLongitude);

var degrees = value.Value[0].ToDouble();
var minutes = value.Value[1].ToDouble();
var seconds = value.Value[2].ToDouble();
minutes += 60 * (degrees - Math.Floor(degrees));
degrees = Math.Floor(degrees);
seconds += 60 * (minutes - Math.Floor(minutes));
minutes = Math.Floor(minutes);
var pos = $"{degrees} {minutes} {seconds}";
```

Did not check how to do negative values based on the `LatitudeRef`

**Reply by Dyloneus** [reply]
The 3rd value in the array (seconds) is always zero (Numerator = 0, Denominator = 1) when using the ios image attached to the original post. 

It's populated for **non-ios** images (such as the one attached to this post) 
![20200812_151254](https://user-images.githubusercontent.com/17324490/93903345-71cf8f00-fcf0-11ea-82df-e9d0ebabe565.jpg)

The issue appears to be the exifProfile.GetValue() method which returns partial data for the rationals used to calculate gps coordinates.

Ive stumbled upon [this interesting post](https://ruk.ca/content/mysterious-case-why-photos-and-preview-couldnt-see-gps-data-embedded-exif-my-jpegs) on the difference between signed and unsigned rationals and the ios stance on the subject. Just wondering if this has anything to do with this issue?

Thanks again

**Reply by dlemstra** [reply]
The seconds are indeed zero in the image but that is not a problem. That is an allowed value and how it is stored in the image.

**Reply by Dyloneus** [reply]
If you download both images and check the gps exif data ( either using https://www.pic2map.com/ or just checking the file properties), you will see that seconds **IS** populated, but magick.net does not (or can not) extract it for the ios image.

That's the problem. Seconds is always zero for ios images.

**Reply by dlemstra** [reply]
Seconds is not populated, it is calculated. As you can see in my example earlier this is done:

`seconds += 60 * (minutes - Math.Floor(minutes));`

And that is why you see a seconds value on that site.

**Reply by Dyloneus** [reply]
Thanks Dirk,

My understanding of how this was calculated was wrong. 

Works perfectly!

Much appreciated.
